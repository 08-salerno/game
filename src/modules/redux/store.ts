import { configureStore } from '@reduxjs/toolkit';
import createSagaMiddleware from 'redux-saga';
import userReducer from './slices/userSlice';
import rootSaga from './sagas/root.saga';

const sagaMiddleware = createSagaMiddleware();

const store = configureStore({
  reducer: {
    user: userReducer,
  },
  middleware: [sagaMiddleware],
});

sagaMiddleware.run(rootSaga);

export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch

export default store;
