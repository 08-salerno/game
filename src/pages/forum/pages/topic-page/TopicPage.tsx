import React, { useEffect, useState } from 'react';
import { useParams } from 'react-router-dom';
import { ForumRouteParams } from '../../routes';
import {
  defaultQueryOffset, getComments, getTopic, leaveComment,
} from '../../api';
import { Topic } from '../../types/topic';
import TopicPreviewer from '../../components/topic-previewer/TopicPreviewer';
import { Comment } from '../../types/comment';
import CommentViewer from './components/comment-viewer/CommentViewer';
import CommentLeaver from './components/comment-leaver/CommentLeaver';

const TopicPage: React.VFC = () => {
  const { topicId } = useParams<ForumRouteParams>();

  const [loading, setLoading] = useState(true);
  const [topic, setTopic] = useState<Topic | null>(null);
  const [loadingComments, setLoadingComments] = useState(true);
  const [comments, setComments] = useState<Comment[]>([]);
  const [commentsOffset, setCommentsOffset] = useState(0);

  useEffect(() => {
    getTopic(topicId)
      .then((topic) => setTopic(topic || null))
      .catch(() => setTopic(null))
      .finally(() => setLoading(false));
  }, []);

  useEffect(() => {
    setLoadingComments(true);
    getComments(topicId, commentsOffset)
      .then((newComments) => setComments([...comments, ...newComments]))
      .catch(() => {
        // todo [sitnik] придумать обработку ошибки
      })
      .finally(() => setLoadingComments(false));
  }, [commentsOffset]);

  const handleLeaveComment = (text: string): Promise<void> => {
    // todo [sitnik] получить текущего пользователя из контекста или тп
    const userId = 1;
    return leaveComment(topicId, text, userId);
  };

  const handleLoadMoreCommentButtonClick = (): void => {
    // todo [sitnik] исправить потенциальную ошибку с некорректным смещением
    setCommentsOffset(commentsOffset + defaultQueryOffset);
  };

  return (
    <>
      {loading ? (
        'Загрузка'
      ) : (
        <>
          {!topic ? (
            <div>Не найдено или редирект</div>
          ) : (
            <>
              <TopicPreviewer
                id={topic.id}
                title={topic.title}
                author={topic.author}
                commentCount={topic.commentCount}
                createdAt={topic.createdAt}
              />
              <CommentLeaver handleLeaveComment={handleLeaveComment} />
              <div>
                {comments.map((comment) => (
                  <CommentViewer
                    key={comment.id}
                    id={comment.id}
                    text={comment.text}
                    author={comment.author}
                    createdAt={comment.createdAt}
                  />
                ))}
                {loadingComments ? (
                  'Загрузка комментариев'
                ) : (
                  <div>
                    <button type="button" onClick={handleLoadMoreCommentButtonClick}>
                      Загрузить ещё
                    </button>
                  </div>
                )}
              </div>
            </>
          )}
        </>
      )}
    </>
  );
};

export default TopicPage;
