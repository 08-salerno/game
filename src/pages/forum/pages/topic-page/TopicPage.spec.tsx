import React from 'react';
import { render } from 'enzyme';
import { Provider } from 'react-redux';
import {
  BrowserRouter as Router,
} from 'react-router-dom';
import store from '../../../../modules/redux/store';
import TopicPage from './TopicPage';

it('TopicPage', () => {
  expect(render(<Provider store={store}>
                  <Router>
                    <TopicPage />
                  </Router>
                </Provider>)).toMatchSnapshot();
});
