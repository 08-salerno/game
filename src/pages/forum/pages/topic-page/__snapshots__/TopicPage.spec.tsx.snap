// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopicPage 1`] = `
LoadedCheerio {
  "0": Node {
    "children": Array [
      Node {
        "data": "Загрузка",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "text",
      },
    ],
    "name": "root",
    "next": null,
    "parent": null,
    "prev": null,
    "type": "root",
  },
  "_root": LoadedCheerio {
    "0": Node {
      "children": Array [
        Node {
          "data": "Загрузка",
          "next": null,
          "parent": [Circular],
          "prev": null,
          "type": "text",
        },
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;
